src/components/application_manager/src/application_data_impl.cc

```diff
@@ -582,6 +582,134 @@ void DynamicApplicationDataImpl::set_display_layout(const std::string& layout) {
+             keyboard != (*configurable_keyboards).end();) {
+          auto found_layout = supported_layouts.find(
+              static_cast<mobile_apis::KeyboardLayout::eType>(
+                  (*keyboard)[hmi_request::keyboard_layout].asInt()));
+          if (supported_layouts.end() == found_layout) {
+            keyboard = (*configurable_keyboards).erase(keyboard);
+          } else {
+            ++keyboard;
+          }
+        }
+      };
+
+  auto fill_customizable_layouts_set =
+      [](const smart_objects::SmartObject& keyboard_capabilities,
+         std::set<mobile_apis::KeyboardLayout::eType>& customizable_layouts) {
+        if (!keyboard_capabilities.keyExists(hmi_response::configurable_keys)) {
+          SDL_LOG_WARN(
+              "No configurable keys param in received keyboard capabilities");
+          return;
+        }
+
+        auto configurable_keyboards =
+            keyboard_capabilities[hmi_response::configurable_keys].asArray();
+        if (!configurable_keyboards) {
+          SDL_LOG_WARN("Pointer to configurable_keyboards array is null");
+          return;
+        }
+
+        std::transform(
+            (*configurable_keyboards).begin(),
+            (*configurable_keyboards).end(),
+            std::inserter(customizable_layouts, customizable_layouts.end()),
+            [](const smart_objects::SmartObject& supported_keyboard) {
+              return static_cast<mobile_apis::KeyboardLayout::eType>(
+                  supported_keyboard[hmi_request::keyboard_layout].asInt());
+            });
+      };
+
+  auto add_missing_keyboard_layouts =
+      [](smart_objects::SmartObject& keyboard_capabilities,
+         const std::set<mobile_apis::KeyboardLayout::eType>& supported_layouts,
+         const std::set<mobile_apis::KeyboardLayout::eType>&
+             customizable_layouts) {
+        for (auto layout : supported_layouts) {
+          auto found_layout = customizable_layouts.find(layout);
+
+          if (customizable_layouts.end() == found_layout) {
+            smart_objects::SmartObject default_capabilities(
+                smart_objects::SmartType_Map);
+            default_capabilities[hmi_response::num_configurable_keys] = 0;
+            default_capabilities[hmi_request::keyboard_layout] = layout;
+
+            if (!keyboard_capabilities.keyExists(
+                    hmi_response::configurable_keys)) {
+              auto missing_keyboards =
+                  smart_objects::SmartObject(smart_objects::SmartType_Array);
+              missing_keyboards[0] = default_capabilities;
+              keyboard_capabilities[hmi_response::configurable_keys] =
+                  missing_keyboards;
+            } else {
+              auto keys_array =
+                  keyboard_capabilities[hmi_response::configurable_keys]
+                      .asArray();
+              keys_array->push_back(default_capabilities);
+            }
+          }
+        }
+      };
```

Comment:
@VladSemenyuk @AKalinich-Luxoft @ychernysheva As for me - this function is too long and it difficult to read. 
As a suggestion - we can create three separate functions in this file using the unnamed namespace.

Reply:
But we have previously three separate functions and decided to place them into this function as lambdas. I'm not sure which approach is better, so I propose to left it as is.

Reply:
@LitvinenkoIra @ychernysheva we chose this approach because application class contains a lot of functions for now so replacing with lambdas will reduce the amount of functions and also encapsulates all that tricky logic inside of single function. Also, as for me there are a lot of intersections between the arguments in that function and their lambdas that's why it was done in this way rather than for readability.
P.S. sorry, uncle Bob, but clean code is not for this case :)